File: src/main/java/nars/storage/MemoryOps.java
package nars.storage;

import nars.entity.*;
import nars.io.IInferenceRecorder;
import nars.language.Term;
import nars.main_nogui.ReasonerBatch;

import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

public interface MemoryOps {
    void clear(); /* ---------- access utilities ---------- */

    List<String> getExportStrings();

    IInferenceRecorder getRecorder();

    long getTime();

    boolean noResult();

    Concept nameToConcept(String name);

    Term nameToListedTerm(String name);

    Concept termToConcept(Term term);

    Concept getConcept(Term term);

    float getConceptActivation(Term t);

    void activateConcept(Concept c, BudgetValue b);

    void inputTask(Task task);

    void activatedTask(BudgetValue budget, Sentence sentence, Sentence candidateBelief);

    void derivedTask(Task task);

    void doublePremiseTask(Term newContent, TruthValue newTruth, BudgetValue newBudget);

    void doublePremiseTask(Term newContent, TruthValue newTruth, BudgetValue newBudget, boolean revisible);

    void singlePremiseTask(Term newContent, TruthValue newTruth, BudgetValue newBudget);

    void singlePremiseTask(Term newContent, char punctuation, TruthValue newTruth, BudgetValue newBudget);

    void workCycle(long clock);

    void processNewTask();

    void processNovelTask();

    void processConcept();

    void immediateProcess(Task task);

    void report(Sentence sentence, boolean input);

    AtomicInteger getTaskForgettingRate();

    AtomicInteger getBeliefForgettingRate();

    AtomicInteger getConceptForgettingRate();

    ReasonerBatch getReasoner();

    ConceptBag getConcepts();

    NovelTaskBag getNovelTasks();

    List<Task> getNewTasks();

    Term getCurrentTerm();

    void setCurrentTerm(Term currentTerm);

    Concept getCurrentConcept();

    void setCurrentConcept(Concept currentConcept);

    TaskLink getCurrentTaskLink();

    void setCurrentTaskLink(TaskLink currentTaskLink);

    Task getCurrentTask();

    void setCurrentTask(Task currentTask);

    @org.jetbrains.annotations.Nullable
    TermLink getCurrentBeliefLink();

    void setCurrentBeliefLink(@org.jetbrains.annotations.Nullable TermLink currentBeliefLink);

    @org.jetbrains.annotations.Nullable
    Sentence getCurrentBelief();

    void setCurrentBelief(@org.jetbrains.annotations.Nullable Sentence currentBelief);

    Stamp getNewStamp();

    void setNewStamp(Stamp newStamp);

    Map<Term, Term> getSubstitute();

    void setSubstitute(Map<Term, Term> substitute);

    void setRecorder(IInferenceRecorder recorder);
}

File: src/main/java/nars/storage/MemoryOps.kt
package nars.storage

import nars.entity.*
import nars.io.IInferenceRecorder
import nars.language.Term
import nars.main_nogui.ReasonerBatch
import java.util.concurrent.atomic.AtomicInteger

interface MemoryOps {
    fun clear() /* ---------- access utilities ---------- */
    val exportStrings: List<String?>?
    var recorder: IInferenceRecorder?
    val time: Long
    fun noResult(): Boolean
    fun nameToConcept(name: String?): Concept?
    fun nameToListedTerm(name: String?): Term?
    fun termToConcept(term: Term?): Concept?
    fun getConcept(term: Term?): Concept?
    fun getConceptActivation(t: Term?): Float
    fun activateConcept(c: Concept?, b: BudgetValue?)
    fun inputTask(task: Task?)
    fun activatedTask(budget: BudgetValue?, sentence: Sentence?, candidateBelief: Sentence?)
    fun derivedTask(task: Task?)
    fun doublePremiseTask(newContent: Term?, newTruth: TruthValue?, newBudget: BudgetValue?)
    fun doublePremiseTask(newContent: Term?, newTruth: TruthValue?, newBudget: BudgetValue?, revisible: Boolean)
    fun singlePremiseTask(newContent: Term?, newTruth: TruthValue?, newBudget: BudgetValue?)
    fun singlePremiseTask(newContent: Term?, punctuation: Any, newTruth: TruthValue?, newBudget: BudgetValue?)
    fun workCycle(clock: Long)
    fun processNewTask()
    fun processNovelTask()
    fun processConcept()
    fun immediateProcess(task: Task?)
    fun report(sentence: Sentence?, input: Boolean)
    val taskForgettingRate: AtomicInteger?
    val beliefForgettingRate: AtomicInteger?
    val conceptForgettingRate: AtomicInteger?
    val reasoner: ReasonerBatch?
    val concepts: ConceptBag?
    val novelTasks: NovelTaskBag?
    val newTasks: List<Task?>?
    var currentTerm: Term?
    var currentConcept: Concept?
    var currentTaskLink: TaskLink?
    var currentTask: Task?
    var currentBeliefLink: TermLink?
    var currentBelief: Sentence?
    var newStamp: Stamp?
    fun getSubstitute(): Map<Term?, Term?>?
    fun setSubstitute(substitute: Map<Term?, Term?>?)
}